From b5af0296cb2f060d9427551e07f470c4d7c39b85 Mon Sep 17 00:00:00 2001
From: Marco Martin <notmart@gmail.com>
Date: Thu, 11 May 2017 10:50:57 +0200
Subject: When building menu hyerarchies, parent menus to their containers

Summary:
Wayland requires menu windows to have a "transient parent" as wl_shell
protocol refuses to either position correcty or assign the corrent
flags to popup windows that don't have a parent.
a lot of popup menus in KDE applications don't have any parent
and will look wrong in wayland. this fixes at least some of the occurrences
of popup menus build by kxmlgui

Test Plan:
"Bookmarks" menu in konsole menubar is now correct in
wayland

Reviewers: #plasma, #plasma_on_wayland, #frameworks, davidedmundson

Reviewed By: #plasma, davidedmundson

Subscribers: davidedmundson, plasma-devel, #frameworks

Tags: #plasma, #frameworks

Differential Revision: https://phabricator.kde.org/D5806
---
 src/kxmlguifactory_p.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/kxmlguifactory_p.cpp b/src/kxmlguifactory_p.cpp
index 5481f93..eb28cb2 100644
--- a/src/kxmlguifactory_p.cpp
+++ b/src/kxmlguifactory_p.cpp
@@ -25,6 +25,7 @@
 
 #include <QWidget>
 #include <QDebug>
+#include <QMenu>
 
 #include "debug.h"
 #include <assert.h>
@@ -592,6 +593,9 @@ bool BuildHelper::processActionElement(const QDomElement &e, int idx)
 
     parentNode->container->insertAction(before, action);
 
+    if (action->menu() && !action->menu()->parentWidget()) {
+        action->menu()->setParent(parentNode->container, action->menu()->windowFlags());
+    }
     // save a reference to the plugged action, in order to properly unplug it afterwards.
     containerClient->actions.append(action);
 
-- 
cgit v0.11.2
